src/lsm/segmented_array.zig

```diff
@@ -219,98 +285,83 @@ fn SegmentedArrayType(
-                assert(existing_a_tail.ptr + existing_a_tail.len == existing_b_tail.ptr);
-            }
+            const a_half_pointer = a_pointer[0..a_half];
+            const b_half_pointer = b_pointer[0..b_half];
 
-            mem.copy(T, b_pointer[existing_b_head.len + elements_b.len ..], existing_b_tail);
-            mem.copy(T, b_pointer[existing_b_head.len..], elements_b);
-            mem.copy(T, b_pointer, existing_b_head);
+            // Move part of `a` forwards to make space for elements.
+            copy_backwards(
+                a_half_pointer,
+                b_half_pointer,
+                cursor.relative_index + elements.len,
+                a_pointer[cursor.relative_index..a_count],
+            );
 
-            mem.copyBackwards(
-                T,
-                a_pointer[existing_a_head.len + elements_a.len ..],
-                existing_a_tail,
+            if (a_half < cursor.relative_index) {
+                // Move the part of `a` that is past the half-way point into `b`.
+                mem.copyBackwards(
+                    T,
+                    b_half_pointer,
+                    a_pointer[a_half..cursor.relative_index],
+                );
+            }
+
+            // Move `elements` into `a` and/or `b`.
+            copy_backwards(
+                a_half_pointer,
+                b_half_pointer,
+                cursor.relative_index,
+                elements,
             );
-            mem.copy(T, a_pointer[existing_a_head.len..], elements_a);
 
             array.indexes[b] = array.indexes[a] + a_half;
             array.increment_indexes_after(b, @intCast(u32, elements.len));
         }
 
+        /// Behaves like mem.copyBackwards, but as if `a` and `b` were a single contiguous slice.
+        /// `target` is the destination index within the concatenation of `a` and `b`.
+        fn copy_backwards(
+            a: []T,
+            b: []T,
+            target: usize,
+            source: []const T,
+        ) void {
+            assert(target + source.len <= a.len + b.len);
+            const target_a = a[@minimum(target, a.len)..@minimum(target + source.len, a.len)];
+            const target_b = b[target -| a.len..(target + source.len) -| a.len];
+            assert(target_a.len + target_b.len == source.len);
+            const source_a = source[0..target_a.len];
+            const source_b = source[target_a.len..];
+            if (target_b.ptr != source_b.ptr) {
+                std.mem.copyBackwards(T, target_b, source_b);
+            }
+            if (target_a.ptr != source_a.ptr) {
+                std.mem.copyBackwards(T, target_a, source_a);
+            }
+        }
+
         /// Insert an empty node at index `node`.
         fn insert_empty_node_at(array: *Self, node_pool: *NodePool, node: u32) void {
             assert(array.node_count + 1 < node_count_max);
 
-            assert(node <= array.node_count);
-            if (node < array.node_count) {
```

Comment:
Why do we remove this branch here?

IIRC, there were some surprising edge cases here in general that @ifreund  and I ran into, e.g. where one must be careful with zero-length slices and bounds checking.

Reply:
Oh, like this?

```
const node_count_max = 10;
const nodes: [node_count_max]u8 = undefined;
const node_count: usize = node_count_max - 2;
const node: usize = node_count_max - 1;
std.mem.copyBackwards(
        Node,
        nodes[node + 1 .. node_count + 1],
        nodes[node..node_count],
);
 ```

I should have applied some Chestertons fence here.

I will restore these, but I'll try to get the tests to catch this first.

Reply:
I tried adding:

```
    pub fn verify(array: Self) void {
            assert(array.node_count < node_count_max);
```

The tests never trigger this assert, despite filling the array to element_count_max.

Reply:
Trying to figure out what workload would hit the worst-case layout.

Reply:
So far I've concluded that node_count_max was actually too high - https://github.com/tigerbeetledb/tigerbeetle/pull/196/files#diff-db51efa9fd74f65bf1c4b81b196552444fa1fc82ea032743c3565dcb4fa333e0R91-R109.

I can now hit node_count_max in the tests, but I'm still not triggering the bounds check issue.

Reply:
My example above was kinda silly in that node > node_count. This isn't ever actually the case in the code.

We can slice any index `<= node_count_max`.

For `insert_empty_node_at` we have:

```
            assert(node <= array.node_count);
            assert(array.node_count + 1 <= node_count_max);

            mem.copyBackwards(
                ?*[node_capacity]T,
                array.nodes[node + 1 .. array.node_count + 1],
                array.nodes[node..array.node_count],
            );
```

`node <= array.node_count` and `array.node_count + 1 <= node_count_max` so `node + 1 <= node_count_max`.

For `remove_empty_node` we have:

```
            assert(node < array.node_count);
            mem.copy(
                ?*[node_capacity]T,
                array.nodes[node .. array.node_count - 1],
                array.nodes[node + 1 .. array.node_count],
            );
```

`node < array.node_count` and from the invariants `array.node_count <= node_count_max` so `node + 1 <= array.node_count`.

So I think that in both cases the existing asserts rule out bounds-check issues.

Reply:
Great, thanks for double-checking. Good to work this forwards and backwards and then we get deeper understanding all round.